### PR TITLE
Improve validation of field size limits in events.

### PR DESCRIPTION
**Pull request by reivilibre**
Supersedes: #13710 

Base: `develop`

This is a re-introduction of #13710 with added backwards compatibility to make the upgrade path smoother.

This pull request is commit-by-commit review friendly. 

Original commit schedule, with full messages:

<ol>
<li>

Introduce rejection reason for oversized events 

</li>
<li>

Introduce `strict` field on EventSizeError 

</li>
<li>

Make _check_size_limits use byte limits 

</li>
<li>

Persist events with non-strict event size errors as rejected 

</li>
<li>

Also do that for the /send_join response 

</li>
<li>

Add stricter pass of event size checking 

</li>
<li>

Don't give any leniency in future room versions 

</li>
<li>

strict -> unpersistable 

</li>
</ol>

### CODE REVIEW

File: synapse/event_auth.py [hunk 1]
```diff
@@ -385,36 +385,36 @@ def _check_size_limits(event: "EventBase") -> None:
 
     # Whole PDU check
     if len(encode_canonical_json(event.get_pdu_json())) > MAX_PDU_SIZE:
-        raise EventSizeError("event too large", strict=True)
+        raise EventSizeError("event too large", unpersistable=True)
```
Comment (DMRobertson) [inline synapse/event_auth.py:388]: I think the docstring above here needs `s/strict/unpersistable/g`

File: synapse/api/room_versions.py [hunk 2]
```diff
@@ -58,7 +58,10 @@ class PushRuleRoomFlag:
     EXTENSIBLE_EVENTS = "org.matrix.msc3932.extensible_events"
 
 
-@attr.s(slots=True, frozen=True, auto_attribs=True)
+# eq=False means equality and hashing are done using object identity, which suits
+# our purposes because each RoomVersion is instantiated only once and some of
+# the fields are not hashable.
+@attr.s(slots=True, frozen=True, auto_attribs=True, eq=False)
```
Comment (DMRobertson) [inline synapse/api/room_versions.py:64]: Could also consider changing `msc3931_push_features` to `Tuple[str, ...]` instead---seems less scary to me, but maybe you disagree.

Comment (DMRobertson) [reply synapse/api/room_versions.py:64]: Basically I'm scared that `RoomVersion is instantiated only once` might become False in the future

**Review by DMRobertson — APPROVED** [review]
Other than one nit and one Opinion, seems sane to me. Glad to get this in.

Edit: complement is a flake AFAICS?

File: synapse/event_auth.py (hunk 1 above)
Comment (squahtx) [reply synapse/event_auth.py:388]: The same applies to `EventSizeError.__init__` and other comments elsewhere

File: synapse/api/room_versions.py (hunk 2 above)
Comment (reivilibre) [reply synapse/api/room_versions.py:64]: That's a fair point. Nobody should be mutating that list anyway.

Comment (reivilibre) [reply synapse/api/room_versions.py:64]: > Basically I'm scared that `RoomVersion is instantiated only once` might become False in the future

(I would say it doesn't seem likely given all the room versions are introduced as constants, but you're right that we don't know what the future holds..)

Comment (reivilibre) [reply synapse/api/room_versions.py]: now done

File: synapse/event_auth.py (hunk 1 above)
Comment (reivilibre) [reply synapse/event_auth.py:389]: well spotted, thanks

**Review by DMRobertson — APPROVED** [review]
LGTM!

**Pull request merged**